src/main/kotlin/org/jitsi/nlj/rtp/codec/vp9/Vp9Packet.kt

```diff
@@ -0,0 +1,285 @@
+ * or not its a keyframe and what its spatial layer index is,
+ * so the constructor allows passing in those values if
+ * they're already known.  If they're null, this instance
+ * will do the parsing itself.
+ */
+class Vp9Packet private constructor (
+    buffer: ByteArray,
+    offset: Int,
+    length: Int,
+    isKeyframe: Boolean?,
+    isStartOfFrame: Boolean?,
+    isEndOfFrame: Boolean?,
+    encodingIndex: Int?,
+    pictureId: Int?,
+    TL0PICIDX: Int?
+) : ParsedVideoPacket(buffer, offset, length, encodingIndex) {
+
+    constructor(
+        buffer: ByteArray,
+        offset: Int,
+        length: Int
+    ) : this(buffer, offset, length,
+        isKeyframe = null,
+        isStartOfFrame = null,
+        isEndOfFrame = null,
+        encodingIndex = null,
+        pictureId = null,
+        TL0PICIDX = null
+    )
+
+    override val isKeyframe: Boolean =
+        isKeyframe ?: DePacketizer.VP9PayloadDescriptor.isKeyFrame(this.buffer, payloadOffset, payloadLength)
+
+    override val isStartOfFrame: Boolean =
+        isStartOfFrame ?: DePacketizer.VP9PayloadDescriptor.isStartOfFrame(buffer, payloadOffset, payloadLength)
+
+    override val isEndOfFrame: Boolean =
+        isEndOfFrame ?: DePacketizer.VP9PayloadDescriptor.isEndOfFrame(buffer, payloadOffset, payloadLength)
+
+    override val layerId: Int
+        get() = if (hasLayerIndices) RtpLayerDesc.getIndex(0, spatialLayerIndex, temporalLayerIndex) else super.layerId
+
+    /** End of VP9 picture is the marker bit. Note frame/picture distinction. */
+    /* TODO: not sure this should be the override from [ParsedVideoPacket] */
+    val isEndOfPicture: Boolean
+        /** This uses [get] rather than initialization because [isMarked] is a var. */
+        get() = isMarked
+
+    val hasLayerIndices =
+        DePacketizer.VP9PayloadDescriptor.hasLayerIndices(buffer, payloadOffset, payloadLength)
+
+    val hasPictureId =
+        DePacketizer.VP9PayloadDescriptor.hasPictureId(buffer, payloadOffset, payloadLength)
+
+    val hasExtendedPictureId =
+        DePacketizer.VP9PayloadDescriptor.hasExtendedPictureId(buffer, payloadOffset, payloadLength)
+
+    val hasTL0PICIDX =
+        DePacketizer.VP9PayloadDescriptor.hasTL0PICIDX(buffer, payloadOffset, payloadLength)
+
+    val hasScalabilityStructure =
+        DePacketizer.VP9PayloadDescriptor.hasScalabilityStructure(buffer, payloadOffset, payloadLength)
+
+    val isUpperLevelReference =
+        DePacketizer.VP9PayloadDescriptor.isUpperLevelReference(buffer, payloadOffset, payloadLength)
+
+    val isInterPicturePredicted =
+        DePacketizer.VP9PayloadDescriptor.isInterPicturePredicted(buffer, payloadOffset, payloadLength)
+
+    private var _TL0PICIDX =
+        TL0PICIDX ?: DePacketizer.VP9PayloadDescriptor.getTL0PICIDX(buffer, payloadOffset, payloadLength)
+
+    var TL0PICIDX: Int
+        get() = _TL0PICIDX
+        set(newValue) {
+            _TL0PICIDX = newValue
```

Comment:
Should we not assign this to `_TL0PICIDX` until after we validate we can write it to the buffer?